networkit/cpp/randomization/GlobalCurveballImpl.h

```diff
@@ -0,0 +1,369 @@
+ */
+#ifndef RANDOMIZATION_GLOBAL_CURVEBALL_IMPL_H_
+#define RANDOMIZATION_GLOBAL_CURVEBALL_IMPL_H_
+
+#include <algorithm>
+#include <utility>
+#include <type_traits>
+
+#include "../auxiliary/Log.h"
+#include "../auxiliary/SignalHandling.h"
+#include "../auxiliary/Timer.h"
+#include "../base/Algorithm.h"
+#include "../graph/Graph.h"
+#include "../graph/GraphBuilder.h"
+
+#include "../auxiliary/RadixHeap.h"
+#include "../auxiliary/RandomBipartitionShuffle.h"
+
+#include "GlobalTradeSequence.h"
+#include "GlobalCurveball.h"
+
+namespace NetworKit {
+namespace CurveballDetails {
+
+template <typename T1, typename T2>
+struct PairFirst {
+    T1 operator()(const std::pair<T1, T2>& p) {return p.first;}
+};
+
+/**
+ * Implementation of EM-GCB ("Parallel and I/O-efficient Randomisation of
+ * Massive Networks using Global Curveball Trades", CJ Carstens et al., ESA 2018).
+ * This class should not be used directly but rather through the wrapper class
+ * GlobalCurveball.
+ */
+class GlobalCurveballImpl {
+    using edgelist_type = std::vector<std::pair<node, node> >;
+    using extract_t = PairFirst<node, node>;
+    using tfp_queue_type = Aux::radixheap< std::pair<node, node>, extract_t , node, 256>;
+
+public:
+    GlobalCurveballImpl(const NetworKit::Graph &G) :
+        input_graph_(G)
+    {}
+
+    template <typename TradeSequence>
+    void run(TradeSequence& trade_sequence) {
+        Aux::SignalHandler handler;
+
+        if (hasRun_) {
+            throw std::runtime_error {"Cannot invoke run several times"};
+        }
+
+        Aux::Timer timer;
+        timer.start();
+
+        std::vector<node>
+            neighbourhood_of_u,
+            neighbourhood_of_v,
+            disjoint_neighbours,
+            common_neighbours;
+
+        auto& urng = Aux::Random::getURNG();
+
+        // copy input graph into queue
+        tfp_queue_type current_pq;
+
+        {
+            // Currently we support only undirect graphs, which should however
+            // be easily fixable
+            assert(!input_graph_.isDirected());
+
+            Aux::Timer loadTimer;
+            loadTimer.start();
+
+            input_graph_.forNodes([&](node u) {
+                const auto hashed_u = trade_sequence.hash(u);
+                const auto hint_u = current_pq.get_bucket_key(hashed_u);
+                input_graph_.forNeighborsOf(u, [&](node v) {
+                    if (u > v) return; // only one message per undirected edge
+
+                    const auto hashed_v = trade_sequence.hash(v);
+
+                    if (hashed_u < hashed_v) {
+                        // u is processed before v
+                        current_pq.emplace_in_bucket(hint_u, hashed_u, v);
+                    } else {
+                        current_pq.emplace(hashed_v, hashed_v, u);
+                    }
+                });
+            });
+
+            loadTimer.stop();
+            INFO("Loading graph took ", loadTimer.elapsedMilliseconds(), "ms.");
```

Comment:
Consider turning this into DEBUG. Lots of INFO messages can be annoying e.g. when you run the test suite.